src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -255,21 +303,112 @@ cdef class AioServer:
         return self._server.add_http2_port(address,
                                           server_credentials._credentials)
 
+    async def _server_main_loop(self,
+                                object server_started):
+        self._server.start()
+        cdef RPCState rpc_state
+        server_started.set_result(True)
+
+        while True:
+            # When shutdown begins, no more new connections.
+            if self._status != AIO_SERVER_STATUS_RUNNING:
+                break
+
+            rpc_state = await _server_call_request_call(
+                self._server,
+                self._cq,
+                self._loop)
+
+            self._loop.create_task(_handle_rpc(
+                self._generic_handlers,
+                rpc_state,
+                self._loop))
```

Comment:
An open question, why we should not take care of stopping also the running tasks?

Otherwise, correct me if I'm wrong, we have the risk of stopping the server and stoping finally the program having still Asyncio Tasks in the event loop, which might result on logging in the console information about some destroyed tasks that were never finished [1]

So, having the feeling that besides asking the gRPC library for stopping all RPC tasks, we should take care after of stopping all of the ongoing Asyncio tasks.

[1] https://github.com/python/cpython/blob/master/Lib/asyncio/tasks.py#L168

Reply:
In the shutdown process, it would cancel all gRPC calls (by the end of the grace period). Ongoing RPC will be cancelled, and pending RPC will fail once started. The tricky part here is whether to track pending RPC or not.

For tasks created in gRPC framework on the server-side, if it is not `await`ed by any caller, then the exception won't be propagated to the main thread. Hence, it would just be another entry of log for users application, even if we trace all ongoing RPC, and cancel them explicitly.

On the other hand, tracking all ongoing RPC might slightly drag the performance.

If you feel strongly, I'm okay to change it.

Reply:
So just for have both of us in the same page, for the following snippet stoping the ongoing calls would mean that when the return is executed the server does nothing because the call was canceled?

```python
async def handler(self, request):
    try:
       response = await mydb.create(....)
       await start_batch_process()
       return MessageReponse(
     except grpc.GrpcError as err:
        print("something wrong happened {}".format(err))
        return MessageReponse(created=False)
```

Maybe the value that we would be adding by canceling explicitly the task is too little, but definitely would make the stoping process in a more consistent way.

Happy on leaving it as it is right now and iterate on it if it's needed later.

Reply:
How to inject exception to a running coroutine is a language-level open question. From a framework point of view, IMO we should focus on offering users the ability to do so, e.g. provide `add_done_callback` so they can stop their long-running-batch-process.

Reply:
IMO sooner or later we will end up canceling explicitly all Asyncio tasks. But let's iterate. The current code is a big step forward.

Reply:
Can you elaborate more about that idea? Maybe create an issue to describe it? I'm not sure if I get the full picture here.

Reply:
Sorry for not being so much elaborated in my answers, my mine concern was with the orphan Asyncio task which they could be doing useless work since the request was already canceled under the hood.

True that in an environment where the shutdown was conducted using a grace period you should expect that during that time All Asyncio tasks should finish.

Only in scenarios where there was no grace period or the request took too much time the Asyncio tasks would be left running while the request was already canceled.

As I said, if we eventually need to make the shutdown process more consistent we will always have the chance of closing explicitly the Asyncio tasks by just sending a cancellation event.

Reply:
Now I see what you are suggesting. I have a fix in `_handle_cancellation_from_core`, but I don't have a perfect solution here.

The naive way of handling this is use a list to track every single task created, and explicitly cancel them during shutdown. Personally, I would like to defer using this approach as much as possible, because it requires extensive tracking for RPC life cycle in our wrapper layer and it might impair performance.

Instead, I choose to handle the event propagated from Core. It should handle more variety of scenarios (other than shutdown, but also application cancellation). However, it requires a separate coroutine to await the cancel event. Admittedly, it slows down the server. The cancellation feature is essential for gRPC wrapper, we needs it for the feature parity.